File: backend/src/models/TopupRequest.js
const mongoose = require('mongoose');

const topupRequestSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  walletId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Wallet',
    required: true
  },
  amount: {
    type: mongoose.Schema.Types.Decimal128,
    required: true,
    min: [1000, 'Minimum top-up amount is 1,000 VND'],
    max: [50000000, 'Maximum top-up amount is 50,000,000 VND'],
    set: val => parseFloat(val).toFixed(2),
    get: val => parseFloat(val.toString())
  },
  method: {
    type: String,
    enum: ['MANUAL', 'BANK_TRANSFER', 'QR_PAY', 'CARD'],
    default: 'MANUAL',
    uppercase: true
  },
  note: {
    type: String,
    trim: true,
    maxlength: 200
  },
  status: {
    type: String,
    enum: ['PENDING', 'APPROVED', 'REJECTED', 'CANCELLED'],
    default: 'PENDING',
    uppercase: true
  },
  referenceNumber: {
    type: String,
    required: true,
    unique: true,
    uppercase: true,
    trim: true
  },
  processedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  processedAt: {
    type: Date
  },
  rejectionReason: {
    type: String,
    trim: true,
    maxlength: 200
  }
}, {
  timestamps: true,
  toJSON: { getters: true },
  toObject: { getters: true }
});

topupRequestSchema.pre('validate', function(next) {
  if (!this.referenceNumber) {
    const timestamp = Date.now();
    const random = Math.random().toString(36).substring(2, 8).toUpperCase();
    this.referenceNumber = `TPRQ${timestamp}${random}`;
  }
  next();
});

module.exports = mongoose.model('TopupRequest', topupRequestSchema);

File: mobile-app/nfc_app/lib/features/topup/domain/topup_request.dart
class TopupRequestItem {
  const TopupRequestItem({
    required this.id,
    required this.referenceNumber,
    required this.amount,
    required this.currency,
    required this.status,
    required this.method,
    this.note,
    this.createdAt,
  });

  final String id;
  final String referenceNumber;
  final double amount;
  final String currency;
  final String status;
  final String method;
  final String? note;
  final DateTime? createdAt;

  factory TopupRequestItem.fromJson(Map<String, dynamic> json) {
    return TopupRequestItem(
      id: json['id']?.toString() ?? '',
      referenceNumber: json['referenceNumber']?.toString() ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      currency: json['currency']?.toString() ?? 'VND',
      status: json['status']?.toString() ?? '',
      method: json['method']?.toString() ?? '',
      note: json['note']?.toString(),
      createdAt: json['createdAt'] != null ? DateTime.tryParse(json['createdAt'].toString()) : null,
    );
  }
}
